### DISCUSSION TITLE
如何在 Ant Design / Pro 中优雅地实现一个登录弹框 dialog / modal？

### DISCUSSION THREAD

**Discussion by Blazero** [ant-design/ant-design#35762] [Q&A]
有一个需求：当用户登录过期时，弹出一个框让其重新登录，而不是跳转到登录界面。这样有助于保留用户之前的工作现场：登录完进行比较，如果登录的用户和之前相同就留在当前页面，如果不同再跳转到主页。

显然，在每个界面都加入一段判断逻辑是不切实际的。我的想法是在 umi-request 中加入一个全局的 [response 拦截器（interceptor）](https://pro.ant.design/docs/request#post-response-interceptors-responseinterceptors)，如果 status 是登录过期，就弹出一个登录框。

仿照[文档](https://pro.ant.design/docs/request#post-response-interceptors-responseinterceptors)在 `app.tsx` 中写了拦截器，但由于此处没有任何 React Components，不知道应该把 `<Modal />` 挂载到哪里，只能使用 `Modal.method()` 弹框。但此时发现弹出框无法访问 `@@intialState`，因此获取不到之前的用户信息，也无法在登录后设置新的用户信息。

看了一下[Modal的文档](https://ant.design/components/modal-cn/#FAQ)， FAQ 中提到：

>直接调用 Modal 方法，antd 会通过 `ReactDOM.render` 动态创建新的 React 实体。其 context 与当前代码所在 context 并不相同，因而无法获取 context 信息。
当你需要 context 信息（例如 ConfigProvider 配置的内容）时，可以通过 Modal.useModal 方法会返回 modal 实体以及 contextHolder 节点。将其插入到你需要获取 context 位置即可。

问题是，我的拦截器和 Components 根本不在一个文件中，如何将 contextHolder 放入相应的位置？

后来我又想，能否将 `<Modal />` 挂载到一个全局的 Component 中（如 `<Header />`），这样就可以获取 context 信息。但是，我该如何在拦截器中控制此 Modal 的弹出与关闭？

请问各位对此有何高见？这两条路哪条能够走通？或者有我没想到的更好实现方案？

**Comment by huntdream** [comment]
可以考虑自己写一个hook，把request包括拦截器的逻辑封装在里面

**Reply by Blazero** [reply]
这样不就要把所有调用request的地方改为调用自己的hook吗

**Comment by UndefinedChai** [comment]
写个全局 event listener 单例，请求失败时 dispatch 一下对应的事件。

**Reply by Blazero** [reply]
事件中似乎同样无法访问 `@@intialState`

**Comment by iversonwool** [comment]
我之前做了个类似的需求，但不是说不登陆就完全不能使用系统，有游客功能，所以不能监听接口返回的未登录错误码。你的想法没错，在每个页面都加一段登录判断的逻辑是不切实际的，所以你要将这段逻辑从UI中剥离出来。也就是说只要检测到未登录就弹出未登录的弹框。`直接调用 Modal 方法，antd 会通过 ReactDOM.render 动态创建新的 React 实体。`，我的做法与这个类似。
